// atomicals-js/dist/commands/get-container-dmitem.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.GetContainerDmitemCommand = void 0;
class GetContainerDmitemCommand {
    constructor(electrumApi, containerName, dmitem) {
        this.electrumApi = electrumApi;
        this.containerName = containerName;
        this.dmitem = dmitem;
    }
    run() {
        return __awaiter(this, void 0, void 0, function* () {
            const responseResult = yield this.electrumApi.atomicalsGetContainerDmitem(this.containerName, this.dmitem);
            return {
                success: true,
                data: responseResult.result
            };
        });
    }
}
exports.GetContainerDmitemCommand = GetContainerDmitemCommand;
